### PR TITLE
feature/refactored chunking code

### PR DESCRIPTION
**Pull request by gecBurton**
## Context

There is some scope to make the chunk-clustering code a little simpler, this will make it:
* easier to read
* a tiny bit faster
* easier to test

Also, the `FileChunker` class has been removed in favour of its only method `chunk_file` which is now static.

## Changes proposed in this pull request

1. two functions `compute_embed_dist` and `compute_token_dist` have been broken out of `create_pdist`. 
2. two instances of replacing python with numpy
3. one (very minor) bug fix??

## Things to check

- [ ] I have added any new ENV vars in all deployed environments
- [x] I have tested any code added or changed
- [x] I have run integration tests https://github.com/i-dot-ai/redbox-copilot/actions/runs/9076094432

### CODE REVIEW

File: redbox/parsing/chunk_clustering.py [hunk 1]
```diff
@@ -119,11 +134,11 @@ def create_pdist(token_counts, pair_embed_dist, weight_embed_dist=0.2, use_log=T
     # bigger weight means more importance of the embedding distance
     if np.std(embed_dist) > 0:
         embed_dist = embed_dist / np.std(embed_dist) * weight_embed_dist
-    if np.std(embed_dist) > 0:
+    if np.std(token_dist) > 0:
```
Comment (gecBurton) [inline redbox/parsing/chunk_clustering.py:137]: is this right @lmwilkigov ?

File: redbox/parsing/chunk_clustering.py [hunk 2]
```diff
@@ -79,35 +79,50 @@ def cluster_chunks(
...
+    n = len(token_counts)
 
     # the token count distance between junk and i and j is the size of minimal text segment
     # containing them, i.e. sum of token counts of all the intermediate chunks
     token_dims = np.tri(n + 1, k=0) * np.concatenate([[0], np.array(token_counts)])
 
     # drop diagonal (sizes of individual chunks)
-    drop_ind = [y - x > 1 for x, y in zip(np.triu_indices(n + 1, k=1)[0], np.triu_indices(n + 1, k=1)[1])]
+    a, b = np.triu_indices(n + 1, k=1)
+    drop_ind = b - a > 1
```
Comment (gecBurton) [inline redbox/parsing/chunk_clustering.py:103]: tiny bit faster

File: redbox/parsing/chunk_clustering.py [hunk 3]
```diff
@@ -119,11 +134,11 @@ def create_pdist(token_counts, pair_embed_dist, weight_embed_dist=0.2, use_log=T
...
         embed_dist = embed_dist / np.std(embed_dist) * weight_embed_dist
-    if np.std(embed_dist) > 0:
+    if np.std(token_dist) > 0:
         token_dist = token_dist / np.std(token_dist) * (1 - weight_embed_dist)
 
     # Phase 2: Combine the two distances into one
 
     # the two above distance are combined either using sum or product (i.e. use_log=T)
-    combined_dist = [x + y for x, y in zip(embed_dist, token_dist)]
+    combined_dist = embed_dist + token_dist
```
Comment (gecBurton) [inline redbox/parsing/chunk_clustering.py:143]: tiny bit faster

File: redbox/tests/parsing/test_file_chunker.py [hunk 4]
```diff
@@ -1,12 +1,11 @@
...
+from redbox.parsing.file_chunker import chunk_file
 
 
 @pytest.mark.parametrize("chunk_clustering, expected_chunk_number", [(False, 54), (True, 13)])
 def test_chunk_file(file_belonging_to_alice, env, chunk_clustering, expected_chunk_number):
-    embedding_model = SentenceTransformerDB(env.embedding_model)
-    file_chunker = FileChunker(embedding_model)
-    chunks = file_chunker.chunk_file(file_belonging_to_alice, chunk_clustering)
+    embedding_model = SentenceTransformerDB(env.embedding_model) if chunk_clustering else None
+    chunks = chunk_file(file_belonging_to_alice, clustering_model=embedding_model)
```
[the remark this answers is not in the document]
Comment (gecBurton) [reply redbox/tests/parsing/test_file_chunker.py:10]: yep. good point. I am not qualified to comment-on / edit this. i am going to roll back these changes

File: redbox/parsing/chunk_clustering.py [hunk 5]
```diff
@@ -117,13 +205,14 @@ def create_pdist(token_counts, pair_embed_dist, weight_embed_dist=0.2, use_log=T
     # make the two distances comparable and then scale them using input weight parameter
     # smaller weight means more importance of the token count distance
     # bigger weight means more importance of the embedding distance
-    if np.std(embed_dist) > 0:
-        embed_dist = embed_dist / np.std(embed_dist) * weight_embed_dist
-    if np.std(embed_dist) > 0:
-        token_dist = token_dist / np.std(token_dist) * (1 - weight_embed_dist)
+    if embed_dist_std := np.std(embed_dist) > 0:
```
[the remark this answers is not in the document]
Comment (gecBurton) [reply redbox/parsing/chunk_clustering.py:208]: so this is a safety-catch - we dont want to dive by zero. in reality this is "never" going to happen unless all values of the `embed_dist` are identical, in which case i guess we just dont care about this measure at all?

File: redbox/parsing/chunk_clustering.py [hunk 6]
```diff
@@ -79,36 +79,124 @@ def cluster_chunks(
...
+    # And rewrite from left-to-bottom (to save space in memory)
+    # [ 40,  60, 100,  50,  90,  60]
+    #
+    # N.B. This is equivalent but slower:
+    # [sum(token_counts[i:j+1]) for i in range(n) for j in range(i+1, n)]
+    return token_dist
+
+
+def create_pdist(
+    token_counts: np.array, pair_embed_dist: np.array, weight_embed_dist: float = 0.2, use_log: bool = True
```
[the remark this answers is not in the document]
Comment (gecBurton) [reply redbox/parsing/chunk_clustering.py:185]: [498bd72](https://github.com/i-dot-ai/redbox-copilot/pull/324/commits/498bd723e228212e77046df6fdaeba74e186cb78)

File: redbox/parsing/chunk_clustering.py [hunk 7]
```diff
@@ -79,36 +79,124 @@ def cluster_chunks(
...
+    #                .2      .4
+    #                        .4
+    #
+    # This is rewritten from left-to-bottom (to save space in memory)
+    # [.3, .3, .4, .2, .4, .4]
+    #
+    return embed_dist
+
+
+def compute_token_dist(token_counts: np.array) -> np.array:
```
[the remark this answers is not in the document]
Comment (gecBurton) [reply redbox/parsing/chunk_clustering.py]: [498bd72](https://github.com/i-dot-ai/redbox-copilot/pull/324/commits/498bd723e228212e77046df6fdaeba74e186cb78)

**Comment by lmwilkigov** [comment]
I love the rationalisation, agree with will on keeping embedding_model name instead of clustering for clarity. Approve for now with the things agreed in the thread above agreed.

**Review by lmwilkigov — APPROVED** [review]

File: redbox/tests/parsing/test_file_chunker.py [hunk 8]
```diff
@@ -1,12 +1,11 @@
 import pytest
 
 from redbox.model_db import SentenceTransformerDB
-from redbox.parsing.file_chunker import FileChunker
+from redbox.parsing.file_chunker import chunk_file
```
[the remark this answers is not in the document]
Comment (gecBurton) [reply redbox/tests/parsing/test_file_chunker.py]: [7583819](https://github.com/i-dot-ai/redbox-copilot/pull/324/commits/75838199b903cd1badf08cf48aaa0c601d673d63)

**Pull request merged**